Implemented zpool sync command

alek-p opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
Implemented 'zpool sync' command. This command is similar to the sync command expect it returns after data has hit the main pool instead of just the ZIL as can be the case with 'sync'.

### Description
<!--- Describe your changes in detail -->
This patch adds a new zpool cmd - zpool sync. This command will wait until the currently open TXG has
been committed to the main pool before returning. It can be ran on all pools or just the specified pools.  Manpage was updated as well as tests added to zfs-tests. Most of the functionality was already there so this patch added an IOCTL and a bit of libzfscore code.
Note I've modified lzc_ioctl() to be able to deal with a NULL ioctl name and source nvl.

### Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
This is something we decided to implemented at Datto to facilitate ZIL flushing.
Because of the way ZIL is implemented flushing ZIL contents is useful when you have an encrypted dataset (note ZIL is still encrypted, see #5769 for details) that has absorbed "sensitive" sync write(s).
zpool sync could also be useful in the rare cases of replacing a failing SLOG to make sure all data has been flushed out.
<!--- If it fixes an open issue, please link to the issue here. -->

### How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
unit testing
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
<!--- If your change is a performance enhancement, please provide benchmarks here. -->

### Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Performance enhancement (non-breaking change which improves efficiency)
- [ ] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

### Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the ZFS on Linux code style requirements.
- [x] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [x] I have added tests to cover my changes.
- [ ] All new and existing tests passed.
- [x] All commit messages are properly formatted and contain `Signed-off-by`.
- [ ] Change has been approved by a ZFS on Linux member.

ahrens commented on cmd/zpool/zpool_main.c:
```diff
@@ -2693,6 +2700,50 @@ zpool_do_import(int argc, char **argv)
+ *     -a      Sync all pools
+ *
+ * Sync the given pool(s) or all pools when the -a option is specified.
+ * Without arguements "zpool sync" will sync all pools.
```
The `-a` flag sounds like a no-op.  I'd suggest removing it.

ahrens commented on lib/libzfs_core/libzfs_core.c:
```diff
@@ -126,17 +127,20 @@ lzc_ioctl(zfs_ioc_t ioc, const char *name,
 
-	(void) strlcpy(zc.zc_name, name, sizeof (zc.zc_name));
+	if (name != NULL)
+		(void) strlcpy(zc.zc_name, name, sizeof (zc.zc_name));
```
Just to make sure I understand, `name` is never actually NULL, right?  This is just in case some future ioctl doesn't need to fill in zc_name?

alek-p replied:
That's right, all the currently existing ioctls provide a name. However, if a future ioctl doesn't need to provide a pool name and gives us NULL we'd panic without this check.
Looks like we should be able to support ioctls without a pool name since we have NO_NAME as a valid value for zfs_ioc_namecheck_t .

alek-p replied:
All the currently existing ioctls provide a name so `name` can't be NULL right now. However, if a future ioctl doesn't need to provide a pool name and gives us NULL we'd panic without this check.
Looks like we should be able to support ioctls without a pool name since we have `NO_NAME` as a valid value for `zfs_ioc_namecheck_t`.
This leaves `source` as the other argument that can't be NULL if we follow the president set in `lzc_get_holds()`.

DeHackEd commented on module/zfs/zfs_ioctl.c:
```diff
@@ -5821,6 +5822,44 @@ zfs_ioc_send_space(const char *snapname, nvlist_t *innvl, nvlist_t *outnvl)
+	/* we checked these as part of the ioctl call */
+	ASSERT3S(spa_writeable(spa), ==, B_TRUE);
+	ASSERT3S(spa_suspended(spa), ==, B_FALSE);
+	ASSERT3S(spa_state(spa), ==, POOL_STATE_ACTIVE);
```
I'm not a fan of this. It's incredibly unlikely but a pool could become suspended between the ioctl handler's check and the assertion's check. No locks are held in the interim between these checks.

ahrens commented on module/zfs/zfs_ioctl.c:
```diff
@@ -5821,6 +5822,44 @@ zfs_ioc_send_space(const char *snapname, nvlist_t *innvl, nvlist_t *outnvl)
+
+	/* we don't want a panicking kernel writing to storage */
+	if (unlikely(oops_in_progress))
+		return (EHOSTDOWN);
```
This check belongs somewhere more generic - not in every ioctl that does a txg_wait_synced().  Perhaps zfsdev_ioctl().  (Although it's surprising that any syscalls are allowed when panicking.  Pretty sure that isn't allowed on illumos.)

behlendorf replied:
I don't think this is necessary.  This function isn't doing anything but waiting for the txg to sync which lots of the existing ioctl's do.  Is the intention to avoid getting blocked here in the case of a panic?

ahrens commented on module/zfs/zfs_ioctl.c:
```diff
@@ -5821,6 +5822,44 @@ zfs_ioc_send_space(const char *snapname, nvlist_t *innvl, nvlist_t *outnvl)
+	if ((err = spa_open(pool, &spa, FTAG)) != 0)
+		return (err);
+
+	/* we checked these as part of the ioctl call */
```
That's true, but they could have changed by now.  Better to observe the `failmode` property than to assert it.

ahrens commented on module/zfs/zfs_ioctl.c:
```diff
@@ -5993,6 +6032,10 @@ zfs_ioctl_init(void)
 
+	zfs_ioctl_register("sync", ZFS_IOC_POOL_SYNC,
+	    zfs_ioc_pool_sync, zfs_secpolicy_none, POOL_NAME,
+	    POOL_CHECK_SUSPENDED | POOL_CHECK_READONLY, B_FALSE, B_TRUE);
```
I would recommend `allow_log=B_FALSE`, especially since you're (appropriately) using `zfs_secopolicy_none`.

ahrens commented on tests/runfiles/linux.run:
```diff
@@ -348,6 +348,9 @@ tests = ['zpool_upgrade_001_pos',
     'zpool_upgrade_009_neg']
 
+[tests/functional/cli_root/zpool_sync]
+tests = ['zpool_sync_001_pos', 'zpool_sync_002_neg']
```
I don't see these files in the review.

behlendorf replied:
Test cases appear to be missing from the PR.

We can also vastly improve the test coverage by updating the `sync_pool` helper function in `libtest.shlib` to simply call `zpool sync`.  I'd also suggest removing the other versions of `sync_pool` in `clean_mirror_common.kshlib`, and `txg_sync` in `create-o_ashift.ksh`.

alek-p replied:
I've updated those but my testing rig is broken so let's see what build bot thinks

alek-p replied:
yup forgot to git add em, they're included now

gmelikov commented on module/zfs/zfs_ioctl.c:
```diff
@@ -5821,6 +5822,44 @@ zfs_ioc_send_space(const char *snapname, nvlist_t *innvl, nvlist_t *outnvl)
 
+/*
+ * Sync the currently open TXG to disk for the specified pool.
+ * This is somewhat simular to 'zfs_sync()'.
```
s/simular/similar

rlaager commented:
Can you elaborate on the use case for "sensitive" data?

What's the current behavior of sync(2) and thus sync(1) with ZFS?

behlendorf commented:
> What's the current behavior of sync(2) and thus sync(1) with ZFS?

For all OpenZFS platforms `sync(1)`, `sync(2)`, and `fsync(2)` all guarantee that your data is durable.  Your system may crash and nothing will be lost.  On FreeBSD and Illumos `sync(1)` has the additional side effect of forcing a full TXG to be written to the primary pool.  Under Linux the data is only guaranteed to be stored in the ZIL due to differences in the VFS.

This has lead to some minor annoyances in the ZFS Test Suite where tests were originally written which depended on this side effect of `sync(1)`.  For example, using `sync(1)` to force a txg sync and an update of the quota information.  This gives us an explicit interface to force the same behavior on Linux.

alek-p commented on lib/libzfs_core/libzfs_core.c:
```diff
@@ -126,17 +127,20 @@ lzc_ioctl(zfs_ioc_t ioc, const char *name,
+		packed = fnvlist_pack(source, &size);
+		zc.zc_nvlist_src = (uint64_t)(uintptr_t)packed;
+		zc.zc_nvlist_src_size = size;
+	}
```
Wouldn't we rather support `lzc_ioctl(ZFS_IOC_NEW_HOTNESS, NULL, NULL, NULL)` than wasting memory on an empty nvlist? Hoping you agree so I've changed `lzc_get_holds()` as well.

behlendorf replied:
Yup, this sounds good as long as we're consistent.

alek-p commented:
> Can you elaborate on the use case for "sensitive" data?

Sure we can take this on IRC or email if you'd like.

Thanks for the reviews folks. I think I've addressed all comments either in code or by responding in this PR. Let me know what you think of the updated version.
Important update - we now do a txg_kick() before waiting on the TXG to since we'd like to have the sync happen asap.

alek-p commented:
In the last update, I've just removed the test for "sync -a" from zpool_sync_001_pos.ksh since that's no longer a valid option.

ahrens commented on module/zfs/zfs_ioctl.c:
```diff
@@ -5821,6 +5823,54 @@ zfs_ioc_send_space(const char *snapname, nvlist_t *innvl, nvlist_t *outnvl)
+	} else if (spa->spa_failmode != ZIO_FAILURE_MODE_CONTINUE) {
+		spa_close(spa, FTAG);
+		return (EINVAL);
+	}
```
Can you explain why the above code is needed here, and not in all the other ioctls that do txg_wait_synced()?

alek-p replied:
Originally these were just asserted to document the assumptions in the code but I changed that in response to your  comment

>  Better to observe the failmode property than to assert it.

Perhaps I should just remove this and continue to rely on the IOCTL checks to catch this. I'm not making the window that much smaller here...

ahrens commented on module/zfs/zfs_ioctl.c:
```diff
@@ -5821,6 +5823,54 @@ zfs_ioc_send_space(const char *snapname, nvlist_t *innvl, nvlist_t *outnvl)
+	}
+
+	/* force a TXG sync if there isn't one happening */
+	txg_kick(spa_get_dsl(spa));
```
I don't think this is necessary, can you explain what it accomplishes?

alek-p replied:
My understanind was the without txg_kick() we may end up waiting the full txg timeout. I'll double check this

behlendorf commented on man/man8/zpool.8:
```diff
@@ -2235,6 +2241,19 @@ Specify \fBu\fR for a printed representation of the internal representation of t
+.sp .6
+.RS 4n
+This command will trigger TXG sync(s) and will not return until the data is
+committed to the pool(s).
```
Let's wordsmith this a bit to avoid referring to a txg which in an internal detail, instead let's describe the expected behavior.  The two critical things worth mentioning are:

* Forces all in-core dirty data to be written to the primary pool storage and not the log device.
* Forces an update of administrative information including quota reporting.

behlendorf commented on tests/zfs-tests/include/libtest.shlib:
```diff
@@ -3211,21 +3212,9 @@ function sync_pool #pool
-		fi
-		log_must sleep 2
-	done
+	eval $ZPOOL sync $pool
```
Why `eval` instead of a simple `log_must`?

behlendorf commented on tests/zfs-tests/tests/functional/cli_root/zpool_sync/zpool_sync_001_pos.ksh:
```diff
@@ -0,0 +1,68 @@
+verify_runnable "global"
+
+function get_txg {
+	typeset -i txg=$($ZDB -u $1 | $SED -n 's/^.*txg = \(.*\)$/\1/p')
```
s/$ZDB/zdb
s/$SED/sed

behlendorf commented on tests/zfs-tests/tests/functional/cli_root/zpool_sync/zpool_sync_001_pos.ksh:
```diff
@@ -0,0 +1,68 @@
+while [[ $i -lt ${#args[*]} ]]; do
+	orig_txg=$(get_txg $TESTPOOL)
+	log_must touch /$TESTPOOL/$i
+	log_must $ZPOOL ${args[i]}
```
s/$ZPOOL/zpool

behlendorf commented on tests/zfs-tests/tests/functional/cli_root/zpool_sync/zpool_sync_002_neg.ksh:
```diff
@@ -0,0 +1,44 @@
+
+typeset -i i=0
+while [[ $i -lt ${#args[*]} ]]; do
+	log_mustnot $ZPOOL sync ${args[i]}
```
s/$ZPOOL/zpool

ryao commented on lib/libzfs_core/libzfs_core.c:
```diff
@@ -340,6 +344,12 @@ lzc_exists(const char *dataset)
 }
 
+int
+lzc_sync(const char *pool_name)
```
Given that this is supposed to be a stable interface, I would like to see a `nvlist_t *opts` variable here that is passed to the kernel and ignored.

alek-p replied:
maybe even `nvlist_t **opts`? Not sure why we need to allocate a nvlist though, we can just ignore the argument provided to lzc_sync().

ryao replied:
Force of habit from earlier work. As for doing a `nvlist_t **outnvl` as well, that sounds good.

alek-p commented:
I've made changes based on the latest round of review feedback. Thanks for looking guys.

alek-p commented:
hmmm so I think the test errors is: 
`Warning: TestGroup '/var/lib/buildbot/slaves/zfs/Amazon_2015_09_x86_64_Release__TEST_/build/tests/zfstests/tests/functional/cli_root/zpool_sync' not added to this run. Auxiliary script '/var/lib/buildbot/slaves/zfs/Amazon_2015_09_x86_64_Release__TEST_/build/tests/zfstests/tests/functional/cli_root/zpool_sync/setup' failed verification.`
but the permissions look ok:
`create mode 100755 tests/zfs-tests/tests/functional/cli_root/zpool_sync/cleanup.ksh`
What am I missing?

loli10K commented:
@alek-p this should fix the issue with the test suite:

```diff
diff --git a/configure.ac b/configure.ac
index 2b3957b..5827f43 100644
--- a/configure.ac
+++ b/configure.ac
@@ -224,6 +224,7 @@ AC_CONFIG_FILES([
        tests/zfs-tests/tests/functional/cli_root/zpool_set/Makefile
        tests/zfs-tests/tests/functional/cli_root/zpool_status/Makefile
        tests/zfs-tests/tests/functional/cli_root/zpool_upgrade/Makefile
+       tests/zfs-tests/tests/functional/cli_root/zpool_sync/Makefile
        tests/zfs-tests/tests/functional/cli_user/Makefile
        tests/zfs-tests/tests/functional/cli_user/misc/Makefile
        tests/zfs-tests/tests/functional/cli_user/zfs_list/Makefile
diff --git a/tests/zfs-tests/tests/functional/cli_root/Makefile.am b/tests/zfs-tests/tests/functional/cli_root/Makefile.am
index f2e3fd0..c3ed44f 100644
--- a/tests/zfs-tests/tests/functional/cli_root/Makefile.am
+++ b/tests/zfs-tests/tests/functional/cli_root/Makefile.am
@@ -45,4 +45,5 @@ SUBDIRS = \
        zpool_scrub \
        zpool_set \
        zpool_status \
-       zpool_upgrade
+       zpool_upgrade \
+       zpool_sync
diff --git a/tests/zfs-tests/tests/functional/cli_root/zpool_sync/Makefile.am b/tests/zfs-tests/tests/functional/cli_root/zpool_sync/Makefile.am
index 87eebcd..7f5f67d 100644
--- a/tests/zfs-tests/tests/functional/cli_root/zpool_sync/Makefile.am
+++ b/tests/zfs-tests/tests/functional/cli_root/zpool_sync/Makefile.am
@@ -3,4 +3,4 @@ dist_pkgdata_SCRIPTS = \
        cleanup.ksh \
        setup.ksh \
        zpool_sync_001_pos.ksh \
-       zpool_sync_002_neg.ksh \
+       zpool_sync_002_neg.ksh
```

I also noticed `cli_root/zpool_create/create-o_ashift` is failing quite reliably, which is probably my fault: i was hoping to find the cause but i'm swamped with work this week, sorry.

behlendorf commented:
@alek-p I've added a commit with @loli10K's build system fixes to this PR.  Go ahead it squash it the next time you update this PR.

More interestingly, I looked in to the `create-o_ashift` failures which did expose a real issue we need to sort out.  The way this patch implements `zpool sync` it doesn't force a txg sync when there's no dirty data.  This is why the updated loop in `create-o_ashift` wasn't flushing the expect number of uberblocks and failing.

What do you think about adding a force `-f` flag to `zpool sync` which forces the txg even when nothing is dirty.

alek-p commented:
I didn't think there was a use case to sync a non-dirty pool but it looks like there is. Adding `-f` support is easy since I can txg_kick() when `-f` is present. Hoping to be able to post updated version today. Thanks, @loli10K and @behlendorf.

ahrens commented:
@behlendorf Are you sure that the problem is with `zpool sync`, and not with the test case?

@alek-p Assuming that we need to write out the uberblock for a no-op txg, I don't think that `txg_kick()` will do that.

If we need to write out the uberblock when doing `zpool sync` even if the pool is not dirty (so this will be a no-op txg), would it hurt to make `zpool sync` always do that, thus removing the need for another flag?  i.e. make `zpool sync` dirty something in the pool such that it won't be a no-op txg.

behlendorf commented:
@ahrens right, we're on the same page, `txg_wait_syned()` was working as intended.  It was just that the test case in question depends on new uberblocks being written.  I should have mentioned I tweaked the test case in 9e0fbe3 to dirty the pool and verified the uberblocks were then written as expected.

>  Assuming that we need to write out the uberblock for a no-op txg, I don't think that txg_kick() will do that.

As was mentioned earlier I'm fairly sure we do not need `txg_kick()` at all.  The presence of a txg_sync waiter should cause the txg_sync before the maximum timeout.

> would it hurt to make zpool sync always do that,

@alek-p and I briefly talked about this and the consensus was we'd like both behaviors.  `zpool sync` gives us a way to replicate the `sync(1M)`  behavior from illumos, and `zpool sync -f` is useful for certain test cases like the one mentioned previously.  Unfortunately, it seems we we need to add a flag to provide both.

>  i.e. make zpool sync dirty something in the pool such that it won't be a no-op txg

I think that's definitely the way to go.

ahrens commented:
@behlendorf 
> the test case in question depends on new uberblocks being written

It seems a bit heavy weight to me to add a new command line flag just for this test case.  Can the test case explicitly dirty the pool before doing `zpool sync`?

behlendorf commented:
@ahrens sure that'd absolutely work too, but that seems a bit clunky.  How do you feel about an undocumented flag for testing, say `-t` to force a txg sync?

ahrens commented:
@behlendorf Personally I'd prefer to make the test gross to keep the main code cleaner, but I'm OK either way.

alek-p commented:
sadly it's not as easy to force uberblock update without dirty data as I thought. I'm looking into how to implement this

alek-p commented:
yup found out txg_kick() doesn't help since we're still not writing any updates if there is no dirty data.
One thing I could do to force an uberblock update is dirty the top level vdev:
```
spa_config_enter(spa, SCL_CONFIG, FTAG, RW_WRITER);
vdev_config_dirty(spa->spa_root_vdev);
spa_config_exit(spa, SCL_CONFIG, FTAG);
```
Since we're trying to update the uberblock, dirtying the label seems to make sense but the locking is a bit heavy here.
It would be pretty easy to have the test case updated to generate some dirty data as well - I'm thinking make sure `atime=on` and `touch $TESTPOOL/file`. I'm ok with either option as well.

alek-p commented:
I ended up implementing the force `-f` option since that's the path I started on earlier

alek-p commented:
I missed that `zfs-tests/tests/functional/upgrade/upgrade_userobj_001_pos` (and maybe others) call sync_pool without arguments so this needs another update...

behlendorf commented on module/zfs/zfs_ioctl.c:
```diff
@@ -5821,6 +5823,42 @@ zfs_ioc_send_space(const char *snapname, nvlist_t *innvl, nvlist_t *outnvl)
+	force = fnvlist_lookup_boolean_value(innvl, "force");
+	if (force) {
+		spa_config_enter(spa, SCL_CONFIG, FTAG, RW_WRITER);
+		vdev_config_dirty(spa->spa_root_vdev);
```
In addition to writing uberblocks this will cause the configuration to be written including updating the zpool cache file.  I don't think that's a problem, it could even be useful, but I didn't see this mentioned in any of the comments.

alek-p replied:
ok I'll add a mention of this

behlendorf commented on tests/zfs-tests/include/libtest.shlib:
```diff
@@ -3203,29 +3204,28 @@ function get_objnum
+function sync_pool #pool <force>
 {
 	typeset pool=${1:-$TESTPOOL}
+	typeset force=false
```
Let's simplify this by using the same style shorthand as in the line above.  Default this value to false unless $2 was provided in which case use that.

```
typeset force=${2:-false}
```

behlendorf commented:
The new `upgrade_userobj_001_pos` [test failure](http://build.zfsonlinux.org/builders/CentOS%207.1%20x86_64%20%28TEST%29/builds/5466/steps/shell_9/logs/log) appears to have been caused by this change.  My guess is the lighter weight `zpool sync` makes it possible for use to run the `zfs userspace` check to soon.

alek-p commented:
looks like `sync_pool` in `upgrade_userobj_001_pos` wasn't actually needed so I've replaced it with `sleep 1`. This seems to have fixed the `upgrade_userobj_001_pos` failures
